lib/service/listeners.go

```
@@ -0,0 +1,97 @@
+	return process.registeredListenerAddr(listenerAuthSSH)
+}
+
+// NodeSSHAddr returns the node SSH endpoint, if configured and started.
+func (process *TeleportProcess) NodeSSHAddr() (*utils.NetAddr, error) {
+	return process.registeredListenerAddr(listenerNodeSSH)
+}
+
+// ProxySSHAddr returns the proxy SSH endpoint, if configured and started.
+func (process *TeleportProcess) ProxySSHAddr() (*utils.NetAddr, error) {
+	return process.registeredListenerAddr(listenerProxySSH)
+}
+
+// DiagnosticAddr returns the diagnostic endpoint, if configured and started.
+func (process *TeleportProcess) DiagnosticAddr() (*utils.NetAddr, error) {
+	return process.registeredListenerAddr(listenerDiagnostic)
+}
+
+// ProxyKubeAddr returns the proxy kubernetes endpoint, if configured and
+// started.
+func (process *TeleportProcess) ProxyKubeAddr() (*utils.NetAddr, error) {
+	return process.registeredListenerAddr(listenerProxyKube)
+}
+
+// ProxyWebAddr returns the proxy web interface endpoint, if configured and
+// started.
+func (process *TeleportProcess) ProxyWebAddr() (*utils.NetAddr, error) {
+	addr, err := process.registeredListenerAddr(listenerProxyTunnelAndWeb)
+	if err == nil {
+		return addr, nil
+	}
+	return process.registeredListenerAddr(listenerProxyWeb)
+}
+
+// ProxyTunnelAddr returns the proxy reverse tunnel endpoint, if configured and
+// started.
+func (process *TeleportProcess) ProxyTunnelAddr() (*utils.NetAddr, error) {
+	addr, err := process.registeredListenerAddr(listenerProxyTunnelAndWeb)
+	if err == nil {
+		return addr, nil
+	}
+	return process.registeredListenerAddr(listenerProxyTunnel)
+}
+
+func (process *TeleportProcess) registeredListenerAddr(typ listenerType) (*utils.NetAddr, error) {
+	process.Lock()
+	defer process.Unlock()
+	for _, l := range process.registeredListeners {
+		if l.typ == typ {
+			return utils.ParseAddr(l.listener.Addr().String())
+		}
+	}
+	return nil, trace.NotFound("no registered address for type %q", typ)
+}
```

This will be flaky during graceful restart? I wonder if it's better to return an error here if multiple listeners are found? Might even uncover some bugs where things are working with incorrect behavior?

---

Good point, added a check for multiple listeners per type.
This method is currently only used in tests, so a strict error is good.